src/main/java/org/truffleruby/core/regexp/RegexpNodes.java

```diff
@@ -459,12 +461,31 @@ private String formatError(DynamicObject string) {
             return StringUtils.format("invalid byte sequence in %s", Layouts.STRING.getRope(string).getEncoding());
         }
 
+        @Specialization(guards = {
+                "isInitialized(regexp)",
+                "isRubyString(string)",
+                "isValidEncoding(string, rangeNode)",
+                "!isRubyEncoding(isCompatibleNode.executeCompatibleQuery(regexp, string))" })
+        protected Object searchRegionIncompatibleEncoding(DynamicObject regexp, DynamicObject string, int start, int end,
+                boolean forward,
+                @Cached ToEncodingNode toEncodingNode,
+                @Cached("toEncodingNode.executeToEncoding(regexp)") Encoding regexpEncoding,
+                @Cached("toEncodingNode.executeToEncoding(string)") Encoding stringEncoding,
+                @Cached EncodingNodes.CompatibleQueryNode isCompatibleNode) {
+            throw new RaiseException(getContext(), coreExceptions().encodingCompatibilityErrorIncompatible(regexpEncoding, stringEncoding, this));
```

Comment:
I found out that we can use `EncodingNodes.CheckEncodingNode` instead and that removes the need for an extra specialization.